src/lake/Lake/Load/Elab.lean

```diff
@@ -76,48 +76,98 @@ def elabConfigFile (pkgDir : FilePath) (lakeOpts : NameMap String)
+  let mut env ← importModulesUsingCache mod.imports leanOpts 1024
+  -- Apply constants (does not go through the kernel, so order is irrelevant)
+  env := mod.constants.foldl addToEnv env
+  /-
+  Apply extension entries (`PersistentEnvExtension.addEntryFn` is pure and
+  does not have access to the whole environment, so no dependency worries
+  here either
+  -/
+  let extDescrs ← persistentEnvExtensionsRef.get
+  let extNameIdx ← mkExtNameMap 0
+  for (extName, ents) in mod.entries do
+    if let some entryIdx := extNameIdx.find? extName then
+      for ent in ents do
+        env := extDescrs[entryIdx]!.addEntry env ent
+  return env
+
+instance : ToJson Hash := ⟨(toJson ·.val)⟩
+instance : FromJson Hash := ⟨((⟨·⟩) <$> fromJson? ·)⟩
+
+structure ConfigLock where
+  leanHash : String
+  configHash : Hash
+  options : NameMap String
+  deriving ToJson, FromJson
+
+/--
+Import the `.olean` for the configuration file if `reconfigure` is not set and
+an up-to-date one exists (i.e., one with matching configuration and on the same
+toolchain). Otherwise, elaborate the configuration and save it to the `.olean`.
+-/
+def importConfigFile (pkgDir lakeDir : FilePath) (lakeOpts : NameMap String)
 (leanOpts := Options.empty) (configFile := pkgDir / defaultConfigFile) (reconfigure := true) : LogIO Environment := do
-  let olean := configFile.withExtension "olean"
-  let useOLean ← id do
-    if reconfigure then return false
-    let .ok oleanMTime ← getMTime olean |>.toBaseIO | return false
-    unless oleanMTime > (← getMTime configFile) do return false
-    let toolchainFile := wsDir / toolchainFileName
-    let .ok toolchainMTime ← getMTime toolchainFile |>.toBaseIO | return true
-    return oleanMTime > toolchainMTime
+  let some configName := FilePath.mk <$> configFile.fileName
+    | error "invalid configuration file name"
+  let olean := lakeDir / configName.withExtension "olean"
+  let lockFile := lakeDir / configName.withExtension "olean.lock"
+  /-
+  Remark: To prevent race conditions between the `.olean` and its lock file
+  (i.e., one process writes a new configuration while another reads an old one
+  and vice versa), Lake takes opens a single handle on the lock file and
+  acquires a lock on it. The lock is held while the lock data is read and
+  the olean is either imported or a new one is written with new lock data.
+  -/
+  let configHash ← computeTextFileHash configFile
+  let (h, useOLean, options) ← id do
+    if (← lockFile.pathExists) then
+      let h ← IO.FS.Handle.mk lockFile .readWrite
+      unless (← h.tryLock) do
+        error "another process configured the package first"
+      if reconfigure then
+        return (h, false, lakeOpts)
+      let contents ← h.readToEnd; h.rewind
+      let .ok (lock : ConfigLock) := Json.parse contents >>= fromJson?
+        | error "compiled configuration is invalid; you will need to reconfigure"
+      return (h, lock.leanHash = Lean.githash ∧ lock.configHash = configHash, lock.options)
```

Comment:
Apologies for nitpicking changes I requested, but I find this code curious (i.e. not a merge blocker), why does it stray so much from all other traces in Lake work? Why not just `mix` the hashes together instead of putting them in a JSON? It *is* a file trace in the end, isn't it?

Reply:
@Kha There at two different things at play here:
1. Lake stores the options verbatim so the olean can be created with the same options when the cache is invalidate without requiring the user to provide them again.
2. I avoided build-style trace mixing because our hashing algorithm is not cryptographically secure and could have clashes and I want to avoid another `lake clean` bug as much as possible.

Reply:
Ok, I wasn't sure whether the option persistence feature was more accidental or requested